libclamav/readdb.c

```diff
@@ -312,7 +312,100 @@ cl_error_t cli_sigopts_handler(struct cli_matcher *root, const char *virname, co
     return ret;
 }
 
-#define SUB_TOKENS 4
+/**
+ * @brief Parse a regex term: a logical subsignature or yara regex string
+ *
+ * expected format => ^offset:trigger/regex/[cflags]$
+ *
+ * @param root      The matcher root (engine structure containing loaded signature patterns for matching)
+ * @param virname   Name of signature that this regex subsig came from.
+ * @param hexsig    The string containing the regex
+ * @param offset    The string offset where the pattern starts
+ * @param lsigid    An array of 2 uint32_t numbers: lsig_id and subsig_id. May be NULL for testing.
+ * @param options   Database options.  See CL_DB_* macros in clamav.h.
+ * @return cl_error_t
+ */
+static cl_error_t readdb_load_regex_subsignature(struct cli_matcher *root, const char *virname, char *hexsig,
+                                                 const char *offset, const uint32_t *lsigid, unsigned int options)
+{
+    cl_error_t status = CL_EPARSE;
+    cl_error_t ret;
+    char *hexcpy = NULL;
+    char *start  = NULL;
+    char *end    = NULL;
+
+    const char *trigger, *pattern, *cflags;
+    int subtokens_count;
+// The maximum number of `:` delimited fields in a regex subsignature.
+#define MAX_REGEX_SUB_TOKENS 4
```

Comment:
Does this belong at the top of the file?  I don't usually see constants defined within functions.

I generally think of constants as "things that might change later, but not without a revision," so grouping them at the top along with all the other potentially-changing things is a practice I've been fairly consistent with (except for unit tests, where constants are often test data).

Reply:
I would be more inclined to remove the constant and hard-code it.  Using a constant makes it easier to read later.  Putting it at the top doesn't feel right to me.  I actually tried with it at the top at first, and then changed my mind and moved it inside the functions.

Reply:
To elaborate, the use of a constant here to me feels like a failing of how we tokenize strings in C. I ideally it could handle any number of tokens with an iterator.  So it is an undesirable constant, as opposed to something I would want to be configurable given a choice.  If/when we rewrite this I wouldn't preserve it. It is also specific to this function, and no other part of the file.  So I don't think it belongs at the top of the file.